The 1.27 rust-stable bootstrap build step fails on mips64

The error mesage is:
```
`error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-L" "/home/draganm/work/rust/build/mips64el-unknown-linux-gnuabi64/stage0/lib/rustlib/mips64el-unknown-linux-gnuabi64/lib" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.138pesuvab4dy035.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.15kq92zzbmxot4k9.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.16u6js6g0l3k1ic6.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.1ckp3bwk0jmrbky4.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.1im38lueib99jsk0.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.1mvmz58owquyropc.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.1y16o1qfye96o7m0.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.1zeawhkbeobww1zn.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.1zwd8n7bcl3vhvvh.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.23tqyymcb18u96mb.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.2jqywn86b2gsqohu.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.2lyh15q6cjwzy18c.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.3171x0bwu82dptu7.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.35af7odfhiqsblci.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.3ayaeypdcro9d6yk.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.3cx7oljifvb206q7.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.3g94tobrpgum5dk6.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.3rngp6bm2u2q5z0y.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.43v6g0y2xsxoggnt.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.45nf4z58qqykpcpi.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.48721dc4k5qxei0u.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.49a7n47po4ttqjl7.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.4b8ptp1vn215jmoe.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.4dvpkvhelzqn37of.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.4xq48u46a1pwiqn7.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.4ybye971cqflgun6.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.4yh8x2b62dcih00t.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.4ypvbwho0bu5tnww.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.8xzrsc1ux72v29j.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.9elsx31vb4it187.rcgu.o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.e19kx9t7xhzajfl.rcgu.o" "-o" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50" "/home/draganm/work/rust/build/bootstrap/debug/deps/bootstrap-e1de2c0797369e50.crate.allocator.rcgu.o" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs" "-L" "/home/draganm/work/rust/build/bootstrap/debug/deps" "-L" "/home/draganm/work/rust/build/mips64el-unknown-linux-gnuabi64/stage0/lib/rustlib/mips64el-unknown-linux-gnuabi64/lib" "-Wl,-Bstatic" "/home/draganm/work/rust/build/bootstrap/debug/deps/libbootstrap-ae75a7d683cd3583.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/libpetgraph-21e72575365cd228.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/libordermap-5f1d6b4d32742667.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/libfixedbitset-18048728a1801049.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/libtime-736f43167516e58e.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/libtoml-51916d91f90fa53e.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/libnum_cpus-12ac34ed547f3180.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/libgetopts-2a463d2722dba7b8.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/libfiletime-c8a8e6c0f3aa27b0.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/liblibc-2d42ca337778fc07.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/libcfg_if-cf754c2e7fcf3720.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/libcmake-72d576f6c4aa644f.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/libcc-6cc2976f70db579f.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/libserde_json-9db09363bcefa01f.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/libitoa-9993a924e130a429.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/libdtoa-ca37bbda1bb7c5a0.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/libserde-9d7a61a020c48540.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/libnum_traits-77a6d928f7bf0203.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/liblazy_static-83ca103b78acf207.rlib" "/home/draganm/work/rust/build/bootstrap/debug/deps/libbuild_helper-807bedf9f7568fde.rlib" "-Wl,--start-group" "/home/draganm/work/rust/build/mips64el-unknown-linux-gnuabi64/stage0/lib/rustlib/mips64el-unknown-linux-gnuabi64/lib/libstd-8d847bbe97fc9dc5.rlib" "/home/draganm/work/rust/build/mips64el-unknown-linux-gnuabi64/stage0/lib/rustlib/mips64el-unknown-linux-gnuabi64/lib/libpanic_unwind-008c42aad548e7dd.rlib" "/home/draganm/work/rust/build/mips64el-unknown-linux-gnuabi64/stage0/lib/rustlib/mips64el-unknown-linux-gnuabi64/lib/libunwind-0fe53371ec419e32.rlib" "/home/draganm/work/rust/build/mips64el-unknown-linux-gnuabi64/stage0/lib/rustlib/mips64el-unknown-linux-gnuabi64/lib/liballoc_system-9dec1cbd51097ce1.rlib" "/home/draganm/work/rust/build/mips64el-unknown-linux-gnuabi64/stage0/lib/rustlib/mips64el-unknown-linux-gnuabi64/lib/liblibc-6a4fb915dd86d140.rlib" "/home/draganm/work/rust/build/mips64el-unknown-linux-gnuabi64/stage0/lib/rustlib/mips64el-unknown-linux-gnuabi64/lib/liballoc-7ebba6af2d3cc324.rlib" "/home/draganm/work/rust/build/mips64el-unknown-linux-gnuabi64/stage0/lib/rustlib/mips64el-unknown-linux-gnuabi64/lib/libstd_unicode-7a26f8b3cf380464.rlib" "/home/draganm/work/rust/build/mips64el-unknown-linux-gnuabi64/stage0/lib/rustlib/mips64el-unknown-linux-gnuabi64/lib/libcore-b0c2d164a9741309.rlib" "-Wl,--end-group" "/home/draganm/work/rust/build/mips64el-unknown-linux-gnuabi64/stage0/lib/rustlib/mips64el-unknown-linux-gnuabi64/lib/libcompiler_builtins-66d072e25a9acee3.rlib" "-Wl,-Bdynamic" "-l" "util" "-l" "dl" "-l" "rt" "-l" "pthread" "-l" "gcc_s" "-l" "c" "-l" "m" "-l" "rt" "-l" "pthread" "-l" "util"
  = note: /home/draganm/work/rust/build/bootstrap/debug/deps/libbootstrap-ae75a7d683cd3583.rlib(bootstrap-ae75a7d683cd3583.2w6xsa2hbi53xhak.rcgu.o): In function `<std::collections::hash::map::HashMap<K, V, S> as core::clone::Clone>::clone':
          /checkout/src/libstd/collections/hash/map.rs:399:(.text._ZN102_$LT$std..collections..hash..map..HashMap$LT$K$C$$u20$V$C$$u20$S$GT$$u20$as$u20$core..clone..Clone$GT$5clone17hde56aa8496891c67E+0x8c): relocation truncated to fit: R_MIPS_CALL16 against `_Unwind_Resume@@GCC_3.0'
          /home/draganm/work/rust/build/bootstrap/debug/deps/libbootstrap-ae75a7d683cd3583.rlib(bootstrap-ae75a7d683cd3583.2w6xsa2hbi53xhak.rcgu.o): In function `<std::collections::hash::map::HashMap<K, V, S> as core::iter::traits::IntoIterator>::into_iter':
          /checkout/src/libstd/collections/hash/map.rs:1834:(.text._ZN116_$LT$std..collections..hash..map..HashMap$LT$K$C$$u20$V$C$$u20$S$GT$$u20$as$u20$core..iter..traits..IntoIterator$GT$9into_iter17h3971322724732454E+0x84): relocation truncated to fit: R_MIPS_CALL16 against `_Unwind_Resume@@GCC_3.0'
          /home/draganm/work/rust/build/bootstrap/debug/deps/libbootstrap-ae75a7d683cd3583.rlib(bootstrap-ae75a7d683cd3583.2w6xsa2hbi53xhak.rcgu.o): In function `<std::collections::hash::map::HashMap<K, V, S> as core::iter::traits::Extend<(K, V)>>::extend':
          /checkout/src/libstd/collections/hash/map.rs:2581:(.text._ZN136_$LT$std..collections..hash..map..HashMap$LT$K$C$$u20$V$C$$u20$S$GT$$u20$as$u20$core..iter..traits..Extend$LT$$LP$K$C$$u20$V$RP$$GT$$GT$6extend17h8b839c2a084c624aE+0xcc): relocation truncated to fit: R_MIPS_CALL16 against `_Unwind_Resume@@GCC_3.0'
          /home/draganm/work/rust/build/bootstrap/debug/deps/libbootstrap-ae75a7d683cd3583.rlib(bootstrap-ae75a7d683cd3583.2w6xsa2hbi53xhak.rcgu.o): In function `<std::collections::hash::map::HashMap<K, V, S> as core::iter::traits::Extend<(K, V)>>::extend':
          /checkout/src/libstd/collections/hash/map.rs:2581:(.text._ZN136_$LT$std..collections..hash..map..HashMap$LT$K$C$$u20$V$C$$u20$S$GT$$u20$as$u20$core..iter..traits..Extend$LT$$LP$K$C$$u20$V$RP$$GT$$GT$6extend17hb662b5607a17e017E+0xb4): relocation truncated to fit: R_MIPS_CALL16 against `_Unwind_Resume@@GCC_3.0'
          /home/draganm/work/rust/build/bootstrap/debug/deps/libbootstrap-ae75a7d683cd3583.rlib(bootstrap-ae75a7d683cd3583.2w6xsa2hbi53xhak.rcgu.o): In function `<std::collections::hash::map::InternalEntry<K, V, &'a mut std::collections::hash::table::RawTable<K, V>>>::into_entry':
          /checkout/src/libstd/collections/hash/map.rs:1680:(.text._ZN157_$LT$std..collections..hash..map..InternalEntry$LT$K$C$$u20$V$C$$u20$$RF$$u27$a$u20$mut$u20$std..collections..hash..table..RawTable$LT$K$C$$u20$V$GT$$GT$$GT$10into_entry17h96e9d8440f6f2c01E+0x2ac): relocation truncated to fit: R_MIPS_CALL16 against `memcpy@@GLIBC_2.0'
          /checkout/src/libstd/collections/hash/map.rs:1686:(.text._ZN157_$LT$std..collections..hash..map..InternalEntry$LT$K$C$$u20$V$C$$u20$$RF$$u27$a$u20$mut$u20$std..collections..hash..table..RawTable$LT$K$C$$u20$V$GT$$GT$$GT$10into_entry17h96e9d8440f6f2c01E+0x44c): relocation truncated to fit: R_MIPS_CALL16 against `memcpy@@GLIBC_2.0'
          /checkout/src/libstd/collections/hash/map.rs:1686:(.text._ZN157_$LT$std..collections..hash..map..InternalEntry$LT$K$C$$u20$V$C$$u20$$RF$$u27$a$u20$mut$u20$std..collections..hash..table..RawTable$LT$K$C$$u20$V$GT$$GT$$GT$10into_entry17h96e9d8440f6f2c01E+0x478): relocation truncated to fit: R_MIPS_CALL16 against `memcpy@@GLIBC_2.0'
          /home/draganm/work/rust/build/bootstrap/debug/deps/libbootstrap-ae75a7d683cd3583.rlib(bootstrap-ae75a7d683cd3583.2w6xsa2hbi53xhak.rcgu.o): In function `<std::collections::hash::map::InternalEntry<K, V, &'a mut std::collections::hash::table::RawTable<K, V>>>::into_entry':
          /checkout/src/libstd/collections/hash/map.rs:1680:(.text._ZN157_$LT$std..collections..hash..map..InternalEntry$LT$K$C$$u20$V$C$$u20$$RF$$u27$a$u20$mut$u20$std..collections..hash..table..RawTable$LT$K$C$$u20$V$GT$$GT$$GT$10into_entry17he24a1b5fedc0b2f6E+0x26c): relocation truncated to fit: R_MIPS_CALL16 against `memcpy@@GLIBC_2.0'
          /checkout/src/libstd/collections/hash/map.rs:1686:(.text._ZN157_$LT$std..collections..hash..map..InternalEntry$LT$K$C$$u20$V$C$$u20$$RF$$u27$a$u20$mut$u20$std..collections..hash..table..RawTable$LT$K$C$$u20$V$GT$$GT$$GT$10into_entry17he24a1b5fedc0b2f6E+0x46c): relocation truncated to fit: R_MIPS_CALL16 against `memcpy@@GLIBC_2.0'
          /home/draganm/work/rust/build/bootstrap/debug/deps/libbootstrap-ae75a7d683cd3583.rlib(bootstrap-ae75a7d683cd3583.2w6xsa2hbi53xhak.rcgu.o): In function `std::collections::hash::map::robin_hood':
          /checkout/src/libstd/collections/hash/map.rs:503:(.text._ZN3std11collections4hash3map10robin_hood17h012217733b3b4b5aE+0xa4): relocation truncated to fit: R_MIPS_CALL16 against `_Unwind_Resume@@GCC_3.0'
          /home/draganm/work/rust/build/bootstrap/debug/deps/libbootstrap-ae75a7d683cd3583.rlib(bootstrap-ae75a7d683cd3583.2w6xsa2hbi53xhak.rcgu.o): In function `std::collections::hash::map::robin_hood':
          /checkout/src/libstd/collections/hash/map.rs:503:(.text._ZN3std11collections4hash3map10robin_hood17h03200646b79056ceE+0xa4): additional relocation overflows omitted from the output`
```
Not sure yet if there is a bug behind this. Posting this here just in case.

What happens is that we hit the limit of 8K GOT entries per single object file. I've tried using 1.25 instead of 1.26 as bootstrap compiler and the bootstrap build passes with  offending object file having ~7K symbols needing GOT entry. I haven't been able to find exact reason for this. At first glance, while it sound counterintuitive, it seems that we have more inlining on 1.26 which pulls in more symbol references.

I was able to sidestep this by increasing opt-level for bootstrap build to 3. As a stopgap, is it ok to provide and option to bootstrap.py  to build bootstrap tool in release mode ? I can send the PR.

@sanxiyn 
Add -mxgot with cc command ?It seems not usefull ...
export RUSTFLAGS=-Clink-arg=-mxgot  ? It seems not pass to linker ....

```<mock-chroot> sh-4.4# ls `which cc` -l
lrwxrwxrwx 1 root root 3 Jul 20 00:44 /usr/bin/cc -> gcc
<mock-chroot> sh-4.4# cat /usr/bin/cc
#!/bin/bash
exec /usr/bin/gcc.aaa -mxgot -Wa,-xgot $@
<mock-chroot> sh-4.4#```

Output 

```  = note: /usr/bin/ld: /builddir/build/BUILD/rustc-1.27.1-src/build/bootstrap/debug/deps/libbootstrap-b174a1b5f8328193.rlib(bootstrap-b174a1b5f8328193.2w6xsa2hbi53xhak.rcgu.o): in function `<std::collections::hash::map::HashMap<K, V, S> as core::clone::Clone>::clone':
          /builddir/build/BUILD/rustc-1.26.2-src/src/libstd/collections/hash/map.rs:399:(.text._ZN102_$LT$std..collections..hash..map..HashMap$LT$K$C$$u20$V$C$$u20$S$GT$$u20$as$u20$core..clone..Clone$GT$5clone17h55339348c661ea5fE+0x8c): relocation truncated to fit: R_MIPS_CALL16 against `_Unwind_Resume@@GCC_3.0'
          /usr/bin/ld: /builddir/build/BUILD/rustc-1.27.1-src/build/bootstrap/debug/deps/libbootstrap-b174a1b5f8328193.rlib(bootstrap-b174a1b5f8328193.2w6xsa2hbi53xhak.rcgu.o): in function `<std::collections::hash::map::HashMap<K, V, S> as core::iter::traits::IntoIterator>::into_iter':
          /builddir/build/BUILD/rustc-1.26.2-src/src/libstd/collections/hash/map.rs:1834:(.text._ZN116_$LT$std..collections..hash..map..HashMap$LT$K$C$$u20$V$C$$u20$S$GT$$u20$as$u20$core..iter..traits..IntoIterator$GT$9into_iter17h2f4a7ffbcb2b8c5cE+0x84): relocation truncated to fit: R_MIPS_CALL16 against `_Unwind_Resume@@GCC_3.0'
          /usr/bin/ld: /builddir/build/BUILD/rustc-1.27.1-src/build/bootstrap/debug/deps/libbootstrap-b174a1b5f8328193.rlib(bootstrap-b174a1b5f8328193.2w6xsa2hbi53xhak.rcgu.o): in function `<std::collections::hash::map::HashMap<K, V, S> as core::iter::traits::Extend<(K, V)>>::extend':
          /builddir/build/BUILD/rustc-1.26.2-src/src/libstd/collections/hash/map.rs:2581:(.text._ZN136_$LT$std..collections..hash..map..HashMap$LT$K$C$$u20$V$C$$u20$S$GT$$u20$as$u20$core..iter..traits..Extend$LT$$LP$K$C$$u20$V$RP$$GT$$GT$6extend17h76fab61df23764faE+0xcc): relocation truncated to fit: R_MIPS_CALL16 against `_Unwind_Resume@@GCC_3.0'
          /usr/bin/ld: /builddir/build/BUILD/rustc-1.27.1-src/build/bootstrap/debug/deps/libbootstrap-b174a1b5f8328193.rlib(bootstrap-b174a1b5f8328193.2w6xsa2hbi53xhak.rcgu.o): in function `<std::collections::hash::map::HashMap<K, V, S> as core::iter::traits::Extend<(K, V)>>::extend':
          /builddir/build/BUILD/rustc-1.26.2-src/src/libstd/collections/hash/map.rs:2581:(.text._ZN136_$LT$std..collections..hash..map..HashMap$LT$K$C$$u20$V$C$$u20$S$GT$$u20$as$u20$core..iter..traits..Extend$LT$$LP$K$C$$u20$V$RP$$GT$$GT$6extend17h8e5cf8a842d5535fE+0xb4): relocation truncated to fit: R_MIPS_CALL16 against `_Unwind_Resume@@GCC_3.0'
          /usr/bin/ld: /builddir/build/BUILD/rustc-1.27.1-src/build/bootstrap/debug/deps/libbootstrap-b174a1b5f8328193.rlib(bootstrap-b174a1b5f8328193.2w6xsa2hbi53xhak.rcgu.o): in function `<std::collections::hash::map::InternalEntry<K, V, &'a mut std::collections::hash::table::RawTable<K, V>>>::into_entry':
          /builddir/build/BUILD/rustc-1.26.2-src/src/libstd/collections/hash/map.rs:1680:(.text._ZN157_$LT$std..collections..hash..map..InternalEntry$LT$K$C$$u20$V$C$$u20$$RF$$u27$a$u20$mut$u20$std..collections..hash..table..RawTable$LT$K$C$$u20$V$GT$$GT$$GT$10into_entry17h6ff08cd565995aa2E+0x2ac): relocation truncated to fit: R_MIPS_CALL16 against `memcpy@@GLIBC_2.0'
          /usr/bin/ld: /builddir/build/BUILD/rustc-1.26.2-src/src/libstd/collections/hash/map.rs:1686:(.text._ZN157_$LT$std..collections..hash..map..InternalEntry$LT$K$C$$u20$V$C$$u20$$RF$$u27$a$u20$mut$u20$std..collections..hash..table..RawTable$LT$K$C$$u20$V$GT$$GT$$GT$10into_entry17h6ff08cd565995aa2E+0x44c): relocation truncated to fit: R_MIPS_CALL16 against `memcpy@@GLIBC_2.0'
          /usr/bin/ld: /builddir/build/BUILD/rustc-1.26.2-src/src/libstd/collections/hash/map.rs:1686:(.text._ZN157_$LT$std..collections..hash..map..InternalEntry$LT$K$C$$u20$V$C$$u20$$RF$$u27$a$u20$mut$u20$std..collections..hash..table..RawTable$LT$K$C$$u20$V$GT$$GT$$GT$10into_entry17h6ff08cd565995aa2E+0x478): relocation truncated to fit: R_MIPS_CALL16 against `memcpy@@GLIBC_2.0'
          /usr/bin/ld: /builddir/build/BUILD/rustc-1.27.1-src/build/bootstrap/debug/deps/libbootstrap-b174a1b5f8328193.rlib(bootstrap-b174a1b5f8328193.2w6xsa2hbi53xhak.rcgu.o): in function `<std::collections::hash::map::InternalEntry<K, V, &'a mut std::collections::hash::table::RawTable<K, V>>>::into_entry':
          /builddir/build/BUILD/rustc-1.26.2-src/src/libstd/collections/hash/map.rs:1680:(.text._ZN157_$LT$std..collections..hash..map..InternalEntry$LT$K$C$$u20$V$C$$u20$$RF$$u27$a$u20$mut$u20$std..collections..hash..table..RawTable$LT$K$C$$u20$V$GT$$GT$$GT$10into_entry17ha43b4a1717213530E+0x26c): relocation truncated to fit: R_MIPS_CALL16 against `memcpy@@GLIBC_2.0'
          /usr/bin/ld: /builddir/build/BUILD/rustc-1.26.2-src/src/libstd/collections/hash/map.rs:1686:(.text._ZN157_$LT$std..collections..hash..map..InternalEntry$LT$K$C$$u20$V$C$$u20$$RF$$u27$a$u20$mut$u20$std..collections..hash..table..RawTable$LT$K$C$$u20$V$GT$$GT$$GT$10into_entry17ha43b4a1717213530E+0x46c): relocation truncated to fit: R_MIPS_CALL16 against `memcpy@@GLIBC_2.0'
          /usr/bin/ld: /builddir/build/BUILD/rustc-1.27.1-src/build/bootstrap/debug/deps/libbootstrap-b174a1b5f8328193.rlib(bootstrap-b174a1b5f8328193.2w6xsa2hbi53xhak.rcgu.o): in function `std::collections::hash::map::robin_hood':
          /builddir/build/BUILD/rustc-1.26.2-src/src/libstd/collections/hash/map.rs:503:(.text._ZN3std11collections4hash3map10robin_hood17h0114868a0efecd28E+0xb0): relocation truncated to fit: R_MIPS_CALL16 against `_Unwind_Resume@@GCC_3.0'
          /usr/bin/ld: /builddir/build/BUILD/rustc-1.26.2-src/src/libstd/collections/hash/map.rs:519:(.text._ZN3std11collections4hash3map10robin_hood17h0114868a0efecd28E+0x260): additional relocation overflows omitted from the output
          collect2: error: ld returned 1 exit status
```

Did anyone ever investigate this? We are still patching in +xgot on Debian which makes a recently-enabled test (https://github.com/rust-lang/rust/pull/83485) fail:

<details><summary>failures</summary>

~~~~
------------------------------------------

error in revision `mips64`: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-12/bin/FileCheck" "--input-file" "/<<PKGBUILDDIR>>/build/x86_64-unknown-linux-gnu/test/assembly/asm/mips-types.mips64/mips-types.s" "/<<PKGBUILDDIR>>/src/test/assembly/asm/mips-types.rs" "--check-prefixes" "CH>
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
/<<PKGBUILDDIR>>/src/test/assembly/asm/mips-types.rs:95:12: error: mips64: expected string not found in input
// mips64: ld $3, %got_disp(extern_static)
           ^
/<<PKGBUILDDIR>>/build/x86_64-unknown-linux-gnu/test/assembly/asm/mips-types.mips64/mips-types.s:24:2: note: scanning from here
 .set push
 ^
/<<PKGBUILDDIR>>/build/x86_64-unknown-linux-gnu/test/assembly/asm/mips-types.mips64/mips-types.s:29:2: note: possible intended match here
 lui $3, %got_hi(extern_static)
 ^
/<<PKGBUILDDIR>>/src/test/assembly/asm/mips-types.rs:105:12: error: mips64: expected string not found in input
// mips64: ld $3, %got_disp(extern_func)
           ^
/<<PKGBUILDDIR>>/build/x86_64-unknown-linux-gnu/test/assembly/asm/mips-types.mips64/mips-types.s:63:2: note: scanning from here
 .set push
 ^
/<<PKGBUILDDIR>>/build/x86_64-unknown-linux-gnu/test/assembly/asm/mips-types.mips64/mips-types.s:68:2: note: possible intended match here
 lui $3, %got_hi(extern_func)
 ^

Input file: /<<PKGBUILDDIR>>/build/x86_64-unknown-linux-gnu/test/assembly/asm/mips-types.mips64/mips-types.s
Check file: /<<PKGBUILDDIR>>/src/test/assembly/asm/mips-types.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
             .
             .
             .
            19:  .set noreorder
            20:  .set nomacro
            21:  .set noat
            22:  .cfi_def_cfa_offset 0
            23:  #APP
            24:  .set push
check:95'0       X~~~~~~~~ error: no match found
            25:  .set at
check:95'0      ~~~~~~~~
            26:  .set macro
check:95'0      ~~~~~~~~~~~
            27:  .set reorder
check:95'0      ~~~~~~~~~~~~~
            28: 
check:95'0      ~
            29:  lui $3, %got_hi(extern_static)
check:95'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:95'1       ?                              possible intended match
            30:  daddu $3, $3, $gp
check:95'0      ~~~~~~~~~~~~~~~~~~
            31:  ld $3, %got_lo(extern_static)($3)
check:95'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            32:  ld $3, 0($3)
check:95'0      ~~~~~~~~~~~~~
            33: 
check:95'0      ~
            34:  .set pop
check:95'0      ~~~~~~~~~
             .
             .
             .
            58:  .set noreorder
            59:  .set nomacro
            60:  .set noat
            61:  .cfi_def_cfa_offset 0
            62:  #APP
            63:  .set push
check:105'0      X~~~~~~~~ error: no match found
            64:  .set at
check:105'0     ~~~~~~~~
            65:  .set macro
check:105'0     ~~~~~~~~~~~
            66:  .set reorder
check:105'0     ~~~~~~~~~~~~~
            67: 
check:105'0     ~
            68:  lui $3, %got_hi(extern_func)
check:105'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:105'1      ?                            possible intended match
            69:  daddu $3, $3, $gp
check:105'0     ~~~~~~~~~~~~~~~~~~
            70:  ld $3, %got_lo(extern_func)($3)
check:105'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            71:  ld $3, 0($3)
check:105'0     ~~~~~~~~~~~~~
            72: ·
check:105'0     ~
            73:  .set pop
check:105'0     ~~~~~~~~~
             .
             .
             .
>>>>>>
~~~~
</details>

According to @jrtc27 this difference is exactly what is expected when +xgot is enabled, so for now I will disable this test in Debian.

--- a/compiler/rustc_target/src/spec/mips64_unknown_linux_gnuabi64.rs
+++ b/compiler/rustc_target/src/spec/mips64_unknown_linux_gnuabi64.rs
@@ -12,7 +12,7 @@ pub fn target() -> Target {
             endian: Endian::Big,
             // NOTE(mips64r2) matches C toolchain
             cpu: "mips64r2".into(),
-            features: "+mips64r2".into(),
+            features: "+mips64r2,+xgot".into(),
             max_atomic_width: Some(64),
             mcount: "_mcount".into(),
 

--- a/compiler/rustc_target/src/spec/mips64el_unknown_linux_gnuabi64.rs
+++ b/compiler/rustc_target/src/spec/mips64el_unknown_linux_gnuabi64.rs
@@ -10,7 +10,7 @@ pub fn target() -> Target {
             abi: "abi64".into(),
             // NOTE(mips64r2) matches C toolchain
             cpu: "mips64r2".into(),
-            features: "+mips64r2".into(),
+            features: "+mips64r2,+xgot".into(),
             max_atomic_width: Some(64),
             mcount: "_mcount".into(),
 

--- a/src/bootstrap/bootstrap.py
+++ b/src/bootstrap/bootstrap.py
@@ -877,6 +877,12 @@ class RustBuild(object):
 
         # preserve existing RUSTFLAGS
         env.setdefault("RUSTFLAGS", "")
+        # we need to explicitly add +xgot here so that we can successfully bootstrap
+        # a usable stage1 compiler
+        # FIXME: remove this if condition on the next bootstrap bump
+        # cfg(bootstrap)
+        if self.build_triple().startswith('mips'):
+            env["RUSTFLAGS"] += " -Ctarget-feature=+xgot"
         target_features = []
         if self.get_toml("crt-static", build_section) == "true":
             target_features += ["+crt-static"]

--- a/tests/assembly/asm/mips-types.rs
+++ b/tests/assembly/asm/mips-types.rs
@@ -72,7 +72,7 @@ macro_rules! check_reg { ($func:ident, $ty:ty, $reg:tt, $mov:literal) => {
 
 // mips32-LABEL: sym_static_32:
 // mips32: #APP
-// mips32: lw $3, %got(extern_static)
+// mips32: lw $3, %got(extern_static)($gp)
 // mips32: #NO_APP
 #[cfg(mips32)]
 #[no_mangle]
@@ -82,7 +82,7 @@ pub unsafe fn sym_static_32() {
 
 // mips32-LABEL: sym_fn_32:
 // mips32: #APP
-// mips32: lw $3, %got(extern_func)
+// mips32: lw $3, %got(extern_func)($gp)
 // mips32: #NO_APP
 #[cfg(mips32)]
 #[no_mangle]
@@ -92,7 +92,9 @@ pub unsafe fn sym_fn_32() {
 
 // mips64-LABEL: sym_static_64:
 // mips64: #APP
-// mips64: ld $3, %got_disp(extern_static)
+// mips64: lui    $3, %got_hi(extern_static)
+// mips64: daddu  $3, $3, $gp
+// mips64: ld     $3, %got_lo(extern_static)($3)
 // mips64: #NO_APP
 #[cfg(mips64)]
 #[no_mangle]
@@ -102,7 +104,9 @@ pub unsafe fn sym_static_64() {
 
 // mips64-LABEL: sym_fn_64:
 // mips64: #APP
-// mips64: ld $3, %got_disp(extern_func)
+// mips64: lui    $3, %got_hi(extern_func)
+// mips64: daddu  $3, $3, $gp
+// mips64: ld     $3, %got_lo(extern_func)($3)
 // mips64: #NO_APP
 #[cfg(mips64)]
 #[no_mangle]

